### PR TITLE
Change stage directories format

### PR DESCRIPTION
**Pull request by nikandfor**
Please guide me if the code could be done more idiomatic for ruby.

Also not sure if the modes is the right place for functions to add.

### CODE REVIEW

File: lib/models.rb [hunk 1]
```diff
@@ -10,6 +10,10 @@ def initialize(slug:, number:, name:)
     @number = number
     @name = name
   end
+
+  def dir_name
```
Comment (rohitpaulk) [inline lib/models.rb]:
```suggestion
  def solution_dir
```

(similar to `Course#solutions_dir` and `Course#dir`)

File: lib/models.rb [hunk 2]
```diff
@@ -64,6 +68,18 @@ def source_repo_url
   def stages_after(course_stage)
     stages.drop_while { |stage| stage.slug != course_stage.slug }.drop(1)
   end
+
+  def stage_path(language, stage, *path)
+    stage_dir = [
```
Comment (rohitpaulk) [inline lib/models.rb]: We don't need the backward compatibility to work - we can just rely on `stage.solution_dir` directly

File: lib/solution_tester.rb [hunk 3]
```diff
@@ -73,7 +73,7 @@ def dockerfile_path
   end
 
   def solution_code_dir_for_stage(stage)
-    "#{solutions_dir}/#{@language.slug}/#{stage.slug}/code"
+    @course.stage_path(@language, stage, "code")
```
Comment (rohitpaulk) [inline lib/solution_tester.rb]:
```suggestion
    "#{stage.solution_dir}/code"
```

(i.e. no backward compatibility)

File: lib/solution_tester.rb [hunk 4]
```diff
@@ -110,6 +110,6 @@ def run_tests_for_stage(stage)
   end
 
   def solution_exists_for_stage?(stage)
-    File.directory?("#{solutions_dir}/#{@language.slug}/#{stage.slug}")
+    File.directory?(@course.stage_path(@language, stage))
```
Comment (rohitpaulk) [inline lib/solution_tester.rb]:
```suggestion
    File.directory?(stage.solution_dir)
```

(i.e. no backward compatibility)

**Review by rohitpaulk — CHANGES_REQUESTED** [review]
Hey @nikandfor - looking good! We can forego the backward compatibility code here to keep things simple. We can open PRs in the course repositories that use this branch, and merge everything together once we know tests pass everywhere.

**Comment by nikandfor** [comment]
made it simple

**Review by rohitpaulk — APPROVED** [review]
Looks good!

**Comment by nikandfor** [comment]
If I change `.github/workflows/test.yml` in courses repos to use this branch we'll need to return them back after rename PRs are merged. Is it ok?

**Comment by rohitpaulk** [comment]
@nikandfor yep, that works! The only requirement here is that any downtime (i.e. tests not working because repositories are incompatible) should be limited to under an hour or so. Another option would be to merge this in when ready, and then rename the other PRs back to master just before merging, instead of after.

**Comment by nikandfor** [comment]
[git](https://github.com/codecrafters-io/build-your-own-git/pull/17)
sdk: nikandfor/rename_stage_dirs
naming: new
tests: all pass

[grep](https://github.com/codecrafters-io/build-your-own-grep/pull/15)
sdk: main
naming: new
tests: verify_templates_compiled failed, others pass

[redis](https://github.com/codecrafters-io/build-your-own-redis/pull/52)
sdk: nikandfor/rename_stage_dirs
naming: old
tests: all pass

looks like with the new sdk tests pass with or without renaming. If it's not a bug, we can safely merge this PR and merge the others easily not changing workflows.

**Comment by rohitpaulk** [comment]
> looks like with the new sdk tests pass with or without renaming. If it's not a bug, we can safely merge this PR and merge the others easily not changing workflows.

@nikandfor ah, didn't think of this. Let's have the tests fail - any directories that don't match the recommended format should throw a friendly error (I'm assuming the tests pass _because_ it's ignoring the older directories)

**Comment by nikandfor** [comment]
Should it be a separate test or use one of existing?

**Comment by rohitpaulk** [comment]
@nikandfor I think it'd fit neatly within `SolutionTester`, somewhere here: https://github.com/codecrafters-io/course-sdk/blob/73ffd37e527c0a6c59fbd6e8f6159a4d3bb1d6ed/lib/solution_tester.rb#L19. Feel free to create a new test if that's easier though - I'm fine with either 🙂

**Comment by rohitpaulk** [comment]
Whether an existing file or new one, let's have it imported/run within https://github.com/codecrafters-io/course-sdk/blob/main/scripts/test.rb

**Comment by nikandfor** [comment]
Wouldn't it be better to run it as a separate test here?
image

**Comment by rohitpaulk** [comment]
@nikandfor hmm, yeah good idea. There'd still be a single https://github.com/codecrafters-io/course-sdk/blob/main/.github/workflows/test-course-definition.yml workflow file to use, yes?

**Comment by nikandfor** [comment]
Yes, we don't need a separate file, but it would be useful to see two separate results: if tests pass and if we have unexpected folders.

**Comment by rohitpaulk** [comment]
@nikandfor yep, makes sense!

**Comment by nikandfor** [comment]
https://github.com/codecrafters-io/build-your-own-redis/actions/runs/3455873160/jobs/5768217454

File: scripts/validate_stages_naming.sh [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+for lang in `ls "solutions"`; do
+	for dir in `ls "solutions/$lang"`; do
+		if contains "$stages_new" "$dir"; then
+			true
+		elif contains "$stages_old" "$dir"; then
+			echo >&2 "old style naming: $lang/$dir"
+			status=1
+		else
+			echo >&2 "unexpected file: $lang/$dir"
```
Comment (rohitpaulk) [inline scripts/validate_stages_naming.sh]: We can remove the `stages_old` handling in this file to keep things simple (people don't have to know what the "old" format was). 

Let's make a small change to the error message to say what is expected. Something like `unexpected file: go/blah, expected format: <stage_number>-<stage_slug> (ex: "01-init")`.

**Comment by nikandfor** [comment]
All five courses are green now, can we go further?

**Comment by rohitpaulk** [comment]
@nikandfor yep, feel free to start merging 👍

**Comment by nikandfor** [comment]
done, all tests pass

**Pull request merged**